src/emailagent.cpp

```diff
@@ -1188,49 +1195,59 @@ void EmailAgent::removeAction(quint64 actionId)
 void EmailAgent::saveAttachmentToDownloads(const QMailMessageId messageId, const QString &attachmentLocation)
 {
     QString attachmentDownloadFolder = QStandardPaths::writableLocation(QStandardPaths::DownloadLocation) + "/mail_attachments/" + attachmentLocation;
-
     // Message and part structure can be updated during attachment download
     // is safer to reload everything
+    QMailMessagePart attachmentPart;
     QMailMessage message (messageId);
+    bool partFound = false;
     for (uint i = 0; i < message.partCount(); i++) {
         QMailMessagePart sourcePart = message.partAt(i);
-        if (attachmentLocation == sourcePart.location().toString(true)) {
-            QString attachmentPath = attachmentDownloadFolder + "/" + sourcePart.displayName();
-            QFile attachmentFile(attachmentPath);
+        if (findAttachmentPartLocation(sourcePart, attachmentLocation, &attachmentPart)) {
+            partFound = true;
+            break;
+        }
+    }
+    if (partFound) {
+        QString attachmentPath = attachmentDownloadFolder + "/" + attachmentPart.displayName();
+        QFile attachmentFile(attachmentPath);
 
-            if (attachmentFile.exists()) {
-                emit attachmentUrlChanged(attachmentLocation, attachmentDownloadFolder);
+        if (attachmentFile.exists()) {
+            emit attachmentUrlChanged(attachmentLocation, attachmentDownloadFolder);
+            updateAttachmentDowloadStatus(attachmentLocation, Downloaded);
+        } else {
+            QString path = attachmentPart.writeBodyTo(attachmentDownloadFolder);
+            if (!path.isEmpty()) {
+                emit attachmentUrlChanged(attachmentLocation, path);
                 updateAttachmentDowloadStatus(attachmentLocation, Downloaded);
             } else {
-                QString path = sourcePart.writeBodyTo(attachmentDownloadFolder);
-                if (!path.isEmpty()) {
-                    emit attachmentUrlChanged(attachmentLocation, path);
-                    updateAttachmentDowloadStatus(attachmentLocation, Downloaded);
-                } else {
-                    updateAttachmentDowloadStatus(attachmentLocation, FailedToSave);
-                }
+                qCDebug(lcDebug) << "ERROR: Failed to save attachment file to location " << attachmentDownloadFolder;
+                updateAttachmentDowloadStatus(attachmentLocation, FailedToSave);
             }
         }
+    } else {
+        qCDebug(lcDebug) << "ERROR: Can't save attachment, location not found " << attachmentLocation;
     }
 }
 
 void EmailAgent::updateAttachmentDowloadStatus(const QString &attachmentLocation, AttachmentStatus status)
 {
-    if (m_attachmentDownloadQueue.contains(attachmentLocation)) {
```

Comment:
Mention error was here, if part is already downloaded and status didn't change(navigate way from page for example) when user press again attachment part is on disk but not in the download queue anymore.